crates/wasm-mutate/src/lib.rs

```diff
@@ -229,215 +221,172 @@ impl WasmMutate {
-                Payload::ElementSection(_) => { info.elements = Some(Range {start:consumed, end: consumed+chunksize} )},
-                Payload::DataSection(_) => { info.data = Some(Range {start:consumed, end: consumed+chunksize})},
-                Payload::End => {break;},
-                _ => {
-
                 }
+                Payload::TableSection(reader) => {
+                    info.tables = Some(info.raw_sections.len());
+                    info.section(SectionId::Table.into(), reader.range(), input_wasm);
+                }
+                Payload::MemorySection(reader) => {
+                    info.memories = Some(info.raw_sections.len());
+                    info.section(SectionId::Memory.into(), reader.range(), input_wasm);
+                }
+                Payload::GlobalSection(reader) => {
+                    info.globals = Some(info.raw_sections.len());
+                    info.section(SectionId::Global.into(), reader.range(), input_wasm);
+                }
+                Payload::ExportSection(mut reader) => {
+                    info.exports = Some(info.raw_sections.len());
+                    info.section(SectionId::Export.into(), reader.range(), input_wasm);
+                }
+                Payload::StartSection { func: _, range: _ } => {
+                    info.is_start_defined = true;
+                }
+                Payload::ElementSection(reader) => {
+                    info.elements = Some(info.raw_sections.len());
+                    info.section(SectionId::Element.into(), reader.range(), input_wasm);
+                }
+                Payload::DataSection(reader) => {
+                    info.data = Some(info.raw_sections.len());
+                    info.section(SectionId::Data.into(), reader.range(), input_wasm);
+                }
+                Payload::CustomSection {
+                    name: _,
+                    data_offset: _,
+                    data: _,
+                    range,
+                } => {
+                    info.section(SectionId::Custom.into(), range, input_wasm);
+                }
+                Payload::AliasSection(reader) => {
+                    info.section(SectionId::Alias.into(), reader.range(), input_wasm);
+                }
+                Payload::UnknownSection {
+                    id,
+                    contents: _,
+                    range,
+                } => {
+                    info.section(id, range, input_wasm);
+                }
+                Payload::DataCountSection { count: _, range } => {
+                    info.section(SectionId::DataCount.into(), range, input_wasm);
+                }
+                Payload::Version { num: _, range: _ } => {
+                    // bypass
+                }
+                Payload::CodeSectionEntry(_) => {
+                    // This is handled by the code section saving
+                }
```

Comment:
The `skip` will solve our problem now, but we will need the full parsing if we want to provide deeper analysis for the `can_mutate` method of the mutators.

Reply:
Remember how we talked about having multiple layers of mutators, each with their own `can_mutate` check? I imagine the top level (the current level we are discussing) won't need to parse the functions at all, just know *whether* functions exist or not and what their types are, which we get from the functions section instead of the code section. Later, once we've decided we want to mutate the code section, we can have a new layer of mutators with their own `can_mutate` checks that actually parse the functions to see if e.g. there are commutative instructions whose operands we can swap.

Reply:
true :)